core/arch/arm/sm/suspend.S

```diff
@@ -0,0 +1,261 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <arm.h>
+#include <arm32_macros.S>
+#include <arm32_macros_cortex_a9.S>
+#include <kernel/asan.h>
+#include <platform_config.h>
+#include <keep.h>
+#include <kernel/unwind.h>
+
+#define PRRR            (0x034)
+#define NMRR            (0x038)
+
+.section .text
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	stmfd	sp!, {r4 - r11, lr}
+	ldr	r4, =suspend_reg_size
+	mov	r5, sp
+	add	r4, r4, #12
+	sub	sp, sp, r4
+	ldr	r3, =suspend_sp
+	read_mpidr r6
+	/*
+	 * Note:
+	 * We assume max 4 cpus and only 1 cluster.
+	 * each cpu consumes 2 words in suspend_sp space
+	 * the &sleep_save_sp_virt = ((u32)suspend_sp + 4 * cpu_id)
+	 */
+	and	r6, r6, #0x3
+	lsl	r6, r6, #0x3
+	stmfd	sp!, {r0, r1}
+	add	r3, r3, r6
+
+	mov	r2, r5
+	mov	r1, r4
+	add	r0, sp, #8
+	bl	cpu_suspend_save
+	adr	lr, cpu_suspend_panic
+	ldmfd	sp!, {r0, r4}
+	/* Jump to arch specific suspend*/
+	blx	r4
+	/* cpu not suspended */
+	ldr	r4, =suspend_reg_size
+	add	r4, r4, #12
+	add	sp, sp, r4
+	ldmia	sp!, {r4 - r11, lr}
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+	bx	lr
+UNWIND( .fnend)
+END_FUNC cpu_suspend
+
+FUNC cpu_suspend_panic, :
+UNWIND( .fnstart)
+	b	.
+UNWIND( .fnend)
+END_FUNC cpu_suspend_panic
+
+FUNC cpu_v7_suspend, :
+UNWIND(.fnstart)
+	stmfd	sp!, {r4 - r11, lr}
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	/* ip changed, ok?? */
+	ldmfd	sp!, {r4 - r11, ip}
+	bx	ip
+UNWIND( .fnend)
+END_FUNC cpu_v7_suspend
+
+FUNC cpu_do_suspend, :
+UNWIND(.fnstart)
+	stmfd	sp!, {r4 - r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr	r4, =0xc07
+	cmp	r5, r4
+	beq	a7_suspend
+	/*
+	 * TODO: add more Cortex A[x] check?
+	 */
+	read_pcr r4
+	read_diag r5
+	stmia	r0!, {r4 - r5}
+a7_suspend:
+	ldmfd	sp!, {r4 - r5}
+	b	cpu_v7_suspend
+UNWIND( .fnend)
+END_FUNC cpu_do_suspend
+
+FUNC cpu_resume, :
+UNWIND(.fnstart)
+	/* Disable IRQ/FIQ */
+	mrs	r1, cpsr
+	orr	r1, r1, #(CPSR_I | CPSR_F)
+	msr	cpsr, r1
+	/* Assmue max 4 cpus and 1 cluster */
+	read_mpidr r1
+	and	r1, r1, #0x3
+
+	adr	r0, _suspend_sp
+	ldr	r2, [r0]
+	lsl	r1, r1, #0x3
+	add	r2, r2, r1
+	/* save sp phys offset at 0x4 */
+	ldr	r0, [r2, #0x4]
+	ldmia	r0!, {r1, sp}
+	/* Need to use r0!, because cpu_do_resume needs it */
+	ldmia	r0!, {r2}
+	/* Jump to cpu_do_resume */
+	bx	r2
+UNWIND( .fnend)
+END_FUNC cpu_resume
+
+FUNC cpu_resume_after_mmu, :
+UNWIND(.fnstart)
+	mov	r0, #0
+	ldmfd	sp!, {r4 - r11}
+	ldmfd	sp!, {r3}
+	bx	r3
+UNWIND( .fnend)
+END_FUNC cpu_resume_after_mmu
+
+FUNC cpu_resume_mmu, :
+UNWIND(.fnstart)
+	ldr	r3, =cpu_resume_after_mmu
+	isb
+	write_sctlr r0
+	mrc     p15, 0, r0, c0, c0, 0
```

Comment:
Could you cross check this (from here to the 2nd `mov r0, r0` below. Looks useless.